### PR TITLE
fix: Add support of auto suggestion memory variable

### LINKED ISSUE #1807: autosuggestion for memory variables in LG editor
**Is your feature request related to a problem? Please describe.**
When user is authoring in LG editor and typing dialog.xxx, they would like to get the auto suggested info for memory variables.

**Describe the solution you'd like**
Following scenarios need to be taken care.
- User defined
eg. user.name; dialog.api_response, user.address.city
- prebuilt func 
eg. this.value, this.turnCount, turn.dialogevent.value
- luis related
eg. #intent.score, @entity,turn.DialogEvent.Value.Entities.xxx

### PR DESCRIPTION
**Pull request by cosmicshuai**
fixes #1807 

## Description
Support of auto suggestion memory variable. 

memoryVariable

### CODE REVIEW

File: Composer/packages/tools/language-servers/language-generation/resources/memoryVariables.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+    "this" : [{"value": ""}, {"turnCount": ""}],
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/resources/memoryVariables.json]: this structure is a little wired, why not

"this": {
  "value": ""
  "turnCount":""
}

File: Composer/packages/tools/language-servers/language-generation/src/utils.ts [hunk 2]
```diff
@@ -132,3 +134,11 @@ export function updateTemplateInContent(content: string, { name, parameters = []
   const resource = LGParser.parse(content);
   return resource.updateTemplate(name, name, parameters, body).toString();
 }
+
+export function loadMemoryVariavles(path: string): object {
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/src/utils.ts]: Can you work with @lei9444 and @zhixzhan to figure out a path to make sure we can load from "pre-defined structure" + "the structure which indexer discovered"? 

we may not implement indexer right now in this PR to discover memory paths, but we need make sure the interface is there to use, we can't too static like this

File: Composer/packages/tools/language-servers/language-generation/resources/memoryVariables.json (hunk 1 above)
Comment (boydc2014) [reply Composer/packages/tools/language-servers/language-generation/resources/memoryVariables.json]: Also, can you work with @luhan2017 to make sure this list is as rich as possible.

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts [hunk 3]
```diff
@@ -43,6 +45,7 @@ export class LGServer {
   protected readonly documents = new TextDocuments();
   protected readonly pendingValidationRequests = new Map<string, number>();
   protected LGDocuments: LGDocument[] = []; // LG Documents Store
+  private readonly memoryVariables: object;
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/src/LGServer.ts]: this should be able to be updated. 

this PR's goal is figure out a way to dynamic set and update this stuff and auto-suggest based on this data source.

File: Composer/packages/tools/language-servers/language-generation/resources/memoryVariables.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/resources/memoryVariables.json]: the format here is really confusing, can you use some sort of linting tool on json?

File: Composer/packages/tools/language-servers/language-generation/src/utils.ts (hunk 2 above)
Comment (cosmicshuai) [reply Composer/packages/tools/language-servers/language-generation/src/utils.ts]: Currently, pre-defined structure and user-defined structure will be stored separately. And the completion suggestions will base on both of them. 
pre-defined structure is stored in an object, imported from lib.
user-defined structure will be updated by a trigger event. A fileResolver will handle reading the file contains the content.

Comment (boydc2014) [reply Composer/packages/tools/language-servers/language-generation/src/utils.ts]: Let's get sometimes today to discuss this, looks like i was not sure about what the static vs dynamic means.

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts [hunk 5]
```diff
@@ -43,8 +44,13 @@ export class LGServer {
   protected readonly documents = new TextDocuments();
   protected readonly pendingValidationRequests = new Map<string, number>();
   protected LGDocuments: LGDocument[] = [];
+  private memoryVariables: Record<string, any> = {};
 
-  constructor(protected readonly connection: IConnection, protected readonly resolver?: FileResolver) {
+  constructor(
+    protected readonly connection: IConnection,
+    protected readonly resolver?: FileResolver,
+    protected readonly memoryResolver?: FileResolver
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/src/LGServer.ts]: First, this memoryResolver don't have to FileResolver, why this is a "FileResolver"??

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts [hunk 6]
```diff
@@ -84,13 +91,52 @@ export class LGServer {
           this.validate(textDocument);
         }
       }
+
+      if (UpdateUserDefinedMemoryVariablesMethod === method) {
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/src/LGServer.ts]: What this is for?

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts (hunk 5 above)
Comment (boydc2014) [reply Composer/packages/tools/language-servers/language-generation/src/LGServer.ts]: Second, you should also enable the E2E workflow that a set of pre-defined variables through your defined interface here, to prove you interface work.

Comment (boydc2014) [reply Composer/packages/tools/language-servers/language-generation/src/LGServer.ts]: By "focus on interface", i'm not referring to "removing anything besides", the use case of this interface is part of interface design.

File: Composer/packages/tools/language-servers/language-generation/src/LGServer.ts [hunk 7]
```diff
@@ -35,6 +35,7 @@ const { check, indexOne } = lgIndexer;
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/src/LGServer.ts]: Let me restate my expectation, 

first, you created a completion experience based on a fixed data structure.  That's been done. 
then what we just have to do, is replace this fixed data structure with a delegate than passed in constructor, that's it, no more, no less.

File: Composer/packages/tools/language-servers/language-generation/src/utils.ts [hunk 8]
```diff
@@ -125,3 +132,8 @@ export function checkTemplate(template: Template): LGDiagnostic[] {
     return diagnostic.message.includes('does not have an evaluator') === false;
   });
 }
+
+export type MemoryVaribleCompletionResult = {
+  endWithDotFlag: boolean;
```
Comment (boydc2014) [inline Composer/packages/tools/language-servers/language-generation/src/utils.ts]: endWithDot is good enough

and what's this flag used for?

Comment (boydc2014) [reply Composer/packages/tools/language-servers/language-generation/src/utils.ts]: This flag looks like a global state, but here is a local state in the `result`, looks weird

File: Composer/packages/server/src/services/project.ts [hunk 9]
```diff
@@ -38,6 +38,11 @@ export class BotProjectService {
     return BotProjectService.currentBotProject?.files.find(file => file.name === name);
   }
 
+  public static staticMemoryResolver(name: string): string[] | undefined {
+    BotProjectService.initialize();
```
Comment (boydc2014) [inline Composer/packages/server/src/services/project.ts]: you don't have to initialize here

File: Composer/packages/server/src/services/project.ts [hunk 10]
```diff
@@ -38,6 +38,11 @@ export class BotProjectService {
...
+    return ['this.value', 'this.turnCount', 'turn.DialogEvent.value', 'intent.score'];
```
Comment (boydc2014) [inline Composer/packages/server/src/services/project.ts]: Please work with @luhan2017  to enrich this.

**Pull request merged**